<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>ALOLA_Manual_1</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    body {
      background-color: #000000;
      color: #ffffff;
      font-family: 'Catamaran', sans-serif;
      font-size: 1,7vh;
      line-height: 1.6;
      margin: 0;
      padding: 25px;
    }

    h1, h2 {
      color: #ffffff;
      font-family: 'Catamaran', sans-serif;
      font-weight: bold; /* Make <strong> text bold */
    }

    a {
      color: #b1adad;
      text-decoration: none;
    }

    a:hover {
      color : #b60505;
    }

    figure {
      background-color: #b5b2b2; /* grey background */
      padding: 10px; /* Padding around the image */
      border-radius: 10px; /* Rounded corners */
      box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1); /* Box shadow for depth */
      margin-bottom: 20px; /* Spacing below the figure */
      display: flex; /* Use flexbox */
      justify-content: center; /* Center horizontally */
      align-items: center; /* Center vertically */
      flex-direction: column; /* Stack elements vertically */
      max-width: 700px
      
    }
    .figure-container {
      max-width: 800px; /* Set the maximum width in pixels */
      margin: 0 auto; /* Center the container horizontally */
    }
    img {
      display: block; /* Make image a block element */
      max-width: 100%; /* Ensure image does not exceed container width */
      height: auto; /* Maintain aspect ratio */
      border-radius: 8px; /* Rounded corners for the image */
    }

    figcaption {
      text-align: center; /* Center-align caption */
      font-style: italic; /* Italicize caption */
      margin-top: 8px; /* Spacing above the caption */
    }
    
    .header-container {
      background-color: #403e3e; /* Dark gray background color */
      color: #fff; /* Text color (white) */
      padding: 20px; /* Add padding around the content */
      border-radius: 10px; /* Rounded corners */
      text-align: center; /* Center vertically */
    }

    /* Center the header content within the container */
    .header-content {
      margin: 0 auto; /* Center the content horizontally */
      font-weight: bold; /* Make <strong> text bold */
    }
    .content-box {
      background-color: #000000; /* Background color */
      border-radius: 10px; /* Rounded corners */
      justify-content: center; /* Center contents horizontally */
      text-decoration: none;
      font-weight: bold; /* Make <strong> text bold */
      align-content: center;
      width: 20%;
      float: left; /* Float left */
      display: inline-block; /* Change display to inline-block */
    }

    .content-box p,
    .content-box ul,
    .content-box li,
    .content-box h1 {
      color: grey; /* Set text color to grey for all text elements */
    }

    .details {
      float: right; /* Float right */
      width: calc(80% - 20px); /* Width of details */
      display: inline-block; /* Change display to inline-block */
      margin-left: 20px; /* Add margin for spacing between boxes */

    }
   
    p, li {
      margin-bottom: 1em;
    }

   
    @media (max-width: 768px) {
      .content-box, .details {
        width: 100%; /* Make them full-width on smaller screens */
        float: none; /* Remove float */
      }
    }
  </style>
</head>
<body>
  <div class="header-container">
    <div class="header-content">
        <h1>ALOLA Manual</h1>
        <p>ALOLA is a tool to provide a user-friendly interface designed for the visualization and customization of the natural products Biosynthetic Gene Clusters (BGCs) derived from antiSMASH results. Additionally, ALOLA facilitates the exploration and modification of the corresponding natural product structures.</p>
    </div>
</div>
<br>
<div class="content-box">
  <h1>Content</h1>
  <ul>
    <li><a href="#alola_navigation">ALOLA Navigation</a></li>
    <li><a href="#how_to_use_alola">How to use ALOLA</a></li>
    <li><a href="#alola_tool_map">ALOLA Tool Map</a></li>
    <li><a href="#pks_nrps">PKSs and NRPSs</a></li>
    <li><a href="#creating_Ripp_bgc">Creating new RiPP BGC</a></li>
    <li><a href="#terpene_bgc">Terpene BGCs</a></li>
    <li><a href="#alola_output">ALOLA Output</a></li>
  </ul>
</div>
<div class="details">
<h2 id="alola_navigation">ALOLA Navigation</h2>
<p>Upon uploading the input file on ALOLA, you'll be greeted with an overview of the ALOLA website. The following introductory photo serves as a visual guide, marked with numbered arrows, each pointing to a specific section or explorer on ALOLA to enhance user familiarity, providing a seamless experience as you explore and utilize the diverse functionalities within this tool.</p>
<figure>
<img src="1004B7D200006352000037EAC3FCE878D136A740.svg" style="width:6.6201in;height:3.9465in" alt=""/>
</figure>
<p><strong>(1) ALOLA Tool bar:</strong> Providing different features enhancing the capabilities of ALOLA</p>
<p><strong>(2) Identified secondary metabolite regions:</strong> Regions are Visible upon Uploading the input file</p>
<p><strong>(3) BGC explorer:</strong> Exploring BGC of the selected region</p>
<p><strong>(4) Protein explorer:</strong> Exploring the BGC encoded proteins and domains of the selected region</p>
<p><strong>(5) Biosynthetic model explorer:</strong> Customizing and modifying the biosynthetic model</p>
<p><strong>(6) Gene explorer:</strong> Directly access to the gene information and annotation</p>
<p><strong>(7) Structure explorer:</strong> Visualizing the final structure of the biosynthetic model</p>
<br>
<h2 id="how_to_use_alola">How to use ALOLA</h2>
<p>ALOLA uses the regions.js file, generated by AntiSMASH, as input data to create dynamic animations of Biosynthetic Gene Clusters (BGCs).</p>
<p>1- After receiving the genomic analysis results from AntiSMASH, click on the "Download" button, and then select "Download all results" on the AntiSMASH web server.</p>
<p>2- In the downloaded folder of the AntiSMASH results, you can find the regions.js file, which ALOLA utilizes. If you"re using AntiSMASH locally, this file can be found in the results folder.</p>
<p>3- Finally click on "Upload file" on the ALOLA web server and select the regions.js file in the downloaded folder from AntiSMASH results (Fig. 1).</p>
<figure>
<img src="10000000000003C00000021C77C4470983B6280A.jpg" style="width:6.6929in;height:2in" alt=""/><figcaption>Figure 1</figcaption>
</figure>
<br>
<p><strong>Create empty BGC</strong></p>
<p>The "Create Empty BGC" button enables you to generate a PKS or NRPS biosynthetic gene cluster (BGC), which you can customize with your desired modules (Fig. 2). This allows you to observe the final result in the explorers, providing a flexible and interactive way to visualize and analyze your customized gene clusters. Create your BGC by incorporating diverse modules or tailoring enzymes to the spaghetti diagram, as described in the "Biosynthetic Model Explorer" section of the ALOLA manual. </p>
<figure>
<img src="10000000000003C00000021CCBE95DA485D32D22.jpg" style="width:6.5543in;height:2.1083in" alt="" /><figcaption>Figure 2</figcaption>
</figure>
<br>
<p><strong>Load Example</strong></p>
<p>The "Load Example" button allows you to load a sample AntiSMASH output from Saccharopolyspora erythraea NRRL 2338 (Fig. 3). By utilizing this option, you can familiarize yourself with ALOLA's functions and explore its capabilities. This serves as a convenient starting point for users to experience and understand the functionalities of ALOLA.</p>
<figure>
<img src="10000000000003C00000021CDC95EEE42818CB23.jpg" style="width:6.6929in;height:1.9in" alt="" /><figcaption>Figure 3</figcaption>
</figure>
<br>
<p><strong>Refresh structures</strong></p>
<p>The "Refresh Structure" button retrieves updated data, ensuring that the displayed structures and spaghetti diagrams accurately reflect the desired modifications made to the Biosynthetic Gene Cluster (BGC) when the "Real time calculation" is disabled (Fig. 4). Employ this feature to dynamically update and showcase the latest modifications in your visualizations.</p>
<figure>
<img src="10000000000003C00000021CFFAE297A7A65FA08.jpg" style="width:6.6929in;height:1.9618in" alt="" /><figcaption>Figure 4</figcaption>
</figure>
<br>
<p><strong>Reset view</strong></p>
<p>The "Reset View" button restores the view to the initial state, eliminating any modifications made to the biosynthetic model (Fig. 5).</p>
<figure>
<img src="10000000000003C00000021C9F27D773A98C6AE0.jpg" style="width:6.6929in;height:1.9583in" alt="" /><figcaption>Figure 5</figcaption>
</figure>
<br>
<p><strong>Real time calculation</strong></p>
<p>The "Real-Time Calculation" button serves as a toggle to enable live, continuous calculation, allowing immediate visualization of the modifications (Fig. 6). When deactivated, spaghetti diagrams and structures are only updated upon pressing the "Refresh Structures" button. This feature proves particularly beneficial when implementing numerous changes, offering flexibility in managing the recalculation process.</p>
<figure>
<img src="10000000000003C00000021C63C3B37371BE7D35.jpg" style="width:6.6083in;height:1.9138in" alt="" /><figcaption>Figure 6</figcaption>
</figure>
<br>
<h2 id="alola_tool_map">ALOLA Tool Map</h2>
<p>- Upon uploading the AntiSMASH output to ALOLA, the identified secondary metabolite regions are displayed at the top of the homepage (e.g. T1PKS, NRPS, Terpene, Lanthipeptide and etc.). </p>
<p>- Each region is linked to different explorers, such as BGC Explorer, Protein Explorer, Gene Explorer, Structure Explorer, and Biosynthetic Model Explorer. These explorers facilitate in-depth exploration and modification of each region:</p>
<p><strong>BGC Explorer</strong></p>
<p>- After selecting a desired region, you can toggle various genes on/off in the BGC Explorer by clicking on the desired gene to observe their effects on the biosynthetic model. Hovering your mouse over each gene in the BGC Explorer provides access to the gene information and annotations in the Gene Explorer.</p>
<p><strong>Protein Explorer</strong></p>
<p>- Hovering over each domain in the Protein Explorer reveals information about the domain, its location, and the corresponding biosynthetic gene in the Gene Explorer. In the Protein Explorer, you can dynamically alter module architecture by dragging and dropping desired modules, observing their impact on the biosynthetic model. This might be specially useful when parts of the gene cluster are on the reverse strand. You can toggle various protein or domain on/off by clicking on them. Additionally, identifying the gene responsible for the biosynthesis of each domain is possible by holding the mouse over each protein, marking the respective gene in red in the BGC Explorer.</p>
<p><strong>Biosynthetic Model explorer</strong></p>
<p>- In ALOLA, you can view only the biosynthetic domains of BGCs from AntiSMASH results, excluding non-relevant domains. </p>
<p>Modify and customize each building block of the shown spaghetti diagram on the natural product BGC assembly line by hovering the mouse over each domain and select your preferred modification within the displayed range. </p>
<p><strong>Expanding BGC</strong></p>
<p>You can additionally integrate diverse tailoring enzymes or modules into the biosynthetic model (Fig. 7). Hover over the "Tailoring Enzyme" (Fig. 7A) and choose your desired enzyme (Fig. 7B).</p>
<figure>
<img src="10000000000003C00000021CA699EDFE170F850F.jpg" style="width:5.5445in;height:4.8091in" alt="" /><figcaption>Figure 7A</figcaption>
</figure>
<figure>
<img src="10000000000003C00000021C3CACF7918DBB3EC8.jpg" style="width:5.3335in;height:5.1819in" alt="" /><figcaption>Figure 7B</figcaption>
</figure>
<p>- Once you've chosen the tailoring enzyme e.g. C-Methyltransferase, from the available range on ALOLA, you can easily select the region for methylation by hovering over the newly added enzyme with your mouse. Observe the dynamic methylation process and its effects on the final product in real time (Fig. 8).</p>
<figure>
<img src="10000000000003C00000021C177DA234E9E63C55.jpg" style="width:6.2917in;height:4.5835in" alt="" /><figcaption>Figure 8</figcaption>
</figure>
<br>
<p><strong>Splicease reaction</strong></p>
<p>To facilitate the selection of splicing sites using the Splicease tailoring enzyme on the Alola website, users are prompted to follow these steps:</p>
<ol type="1">
<li>After adding the Splicease enzyme from the "Add Tailoring Enzyme" list, click on the newly added Splicease enzyme.</li>
<li>This action will prompt a list of potential modification sites to appear.</li>
<li>The user should then carefully select two atoms for the splicing reaction on the final product.</li>
<li>Ensure that the selected atoms are suitable for the desired splicing process, considering factors such as proximity, chemical reactivity, and structural integrity.</li>
</ol>
<p><strong>Important:</strong> To observe the most recent modifications promptly, ensure the "Real-Time Calculation" button is enabled. This feature allows users to view updates in real-time</p>
<br>
<h2 id="pks_nrps">PKSs and NRPSs</h2>
<p>- Upon selecting a PKS or NRPS between the identified secondary metabolite regions, you can integrate new modules into the BGC and observe the changes on spaghetti diagram by simply opting for the "Add Module" feature (Fig. 9).</p>
<figure>
<img src="10000000000003C00000021C75EC413A937BF16C.jpg" style="width:5.7646in;height:4.8256in" alt="" /><figcaption>Figure 9</figcaption>
</figure>
<p>- Suppose you wish to append a new module to a PKS BGC. Upon selecting PKS (Fig. 10), a dialog box will appear, enabling you to choose the Module Type and Substrate for the upcoming module (Fig. 11).</p>
<figure>
<img src="10000000000003C00000021CA5D501EF94565128.jpg" style="width:6.4945in;height:4.7252in" alt="" /><figcaption>Figure 10</figcaption>
</figure>
<figure>
<img src="10000000000003C00000021C41705BF3A8187E64.jpg" style="width:6.4945in;height:4.6252in" alt="" /><figcaption>Figure 11</figcaption>
</figure>
<p>- By hovering the mouse over the "Module Type" you can effortlessly toggle between the Starter Module, Elongation Module, and Terminator Module options, allowing an intuitive selection (Fig. 12).</p>
<figure>
<img src="10000000000003C00000021CE9C813AA76465929.jpg" style="width:6.6193in;height:4.7717in" alt="" /><figcaption>Figure 12</figcaption>
</figure>
<p>- Hovering over the " Substrate " (Fig. 13) allows you to effortlessly navigate through various starter units that can be employed by PKSs. You can conveniently select your desired starter unit for configuring the new adding module and finally submit your request (Fig. 14).</p>
<figure>
<img src="10000000000003C00000021CBCEC03F986C29B11.jpg" style="width:6.8327in;height:4.7673in" alt="" /><figcaption>Figure 13</figcaption>
</figure>
<figure>
<img src="10000000000003C00000021C0CABE5CD1ABC581B.png" style="width:6.8146in;height:4.7717in" alt="" /><figcaption>Figure 14</figcaption>
</figure>
<p>- You can identify the corresponding gene in the BGC Explorer and Protein Explorer by holding the mouse over each domain (highlighted in red). Finally you are able to save the biosynthetic model in PNG or SVG format with and without tailoring enzymes. </p>
<br>
<p><strong>CAL-Domain (Co-enzyme A Ligase Domain)</strong></p>
<p>- The CAL-Domain (Co-enzyme A Ligase Domain) belongs to class I of the adenylate-forming enzyme superfamily. After uploading the results from antiSMASH, if the CAL domain is present in the BGC and is located anywhere other than the first position in the spaghetti diagram, it cannot add any substrate to the diagram in the biosynthetic model explorer.
To place the CAL domain in the first position, hover the mouse over it to identify the corresponding domain in the protein explorer (the domain will change color to red). Then, drag and drop the domain to the first position. Additionally, by hovering over the CAL domain, you can select from a range of NRPS substrates (different amino acids) to be used as the CAL substrate.</p>
<br>
<p><strong>TE-Domain (Thioesterase Domain)</strong></p>
<p>- The Thioesterase (TE) domain catalyzes linear peptide cyclization and releases the final product. Therefore, it should be placed as the last module in the spaghetti diagram within the biosynthetic model explorer. 
If the TE domain is located as the starting module in the selected region after uploading the antiSMASH results to Alola, you can easily use the "Reverse BGC" button to rearrange the BGC. This will position the TE domain at the end of the BGC and the spaghetti diagram.</p>
<h2 id="creating_Ripp_bgc">Creating new RiPP BGC</h2>
<p>- Generate a new RiPP BGC (Fig. 15) based on the current identified secondary metabolite regions, preferably originating from a RiPP BGC, such as various classes of Lanthipeptides. This can be achieved by selecting a RiPP precursor (Fig. 15-16).</p>
<figure>
<img src="10000000000003C00000021C3F5FBF2FB8CE17F5.jpg" style="width:6.5252in;height:4.6917in" alt="" /><figcaption>Figure 15</figcaption>
</figure>
<figure>
<img src="10000000000003C00000021C95A565D97F6A8D15.jpg" style="width:6.7783in;height:5.1193in" alt="" /><figcaption>Figure 16</figcaption>
</figure>
<p>- Upon choosing the desired RiPP precursor, proceed to select the core peptide from the amino acid sequence associated with the chosen RiPP precursor (Fig. 17 (1)) and submit your request (Fig. 17 (2)). In the absence of a specifically selected core peptide, the last 5 amino acids will be automatically utilized:</p>
<figure>
<img src="10000000000003C00000021CF9ED5F1291C0DCF0.jpg" title="fig:" style="width:6.9417in;height:5.2402in" alt="" /><figcaption>Figure 17</figcaption>
</figure>
<p>- Ultimately, you can observe the end product on the assembly line. Integrate diverse tailoring enzymes into the biosynthetic model to see the modifications to the end product (Fig. 7). Hover over the " Tailoring Enzyme" (Fig. 7A) and choose your desired enzyme (Fig. 7B).</p>
<br>
<h2 id="terpene_bgc">Terpene BGCs</h2>
<p><br />
If the identified secondary metabolite regions involve Terpene, upon selecting the Terpene region, a dialogue box will prompt you to first choose a substrate, such as geranyl-/farnesyl-pyrophosphate, squalene, phytoene, etc. This selection enables the processing of the Biosynthetic Gene Cluster (BGC) as a Terpene BGC, providing you with the capability to explore and modify it. Following the substrate selection, you can submit your request and view the results (Fig. 18).</p>
<figure>
<img src="10000000000003C00000021CF08DB9144D0A388D.jpg" style="width:6.7339in;height:5.2291in" alt="" /><figcaption>Figure 18</figcaption>
</figure>
<br>
<p><strong>Important:</strong> After selecting the desired substrate for the cyclization process, users must specify two atoms or desired modification sites:</p>
<ol type="1">
<li>Upon hovering the mouse over the "Cycl" symbol, a list of possible modification sites for the cyclization process will be displayed. </li>
<li><p>Once these sites are chosen, clicking the "Refresh Structure" button will visualize the cyclized product, showcasing the alterations made at the selected modification sites. </p></li>
</ol>
<p>This iterative process enables users to visualize and fine-tune the cyclization reactions with precision, enhancing their ability to explore and modify terpene biosynthetic pathways effectively.</p>
<p><strong>Gene Explorer</strong></p>
<p>- Access detailed gene information and annotations in the Gene Explorer by holding the mouse over each gene in the BGC Explorer or each domain in the Protein Explorer.</p>
<p>- This iterative process enables users to visualize and fine-tune the cyclization reactions with precision, enhancing their ability to explore and modify terpene biosynthetic pathways effectively.</p>
<p><strong>Structure Explorer</strong></p>
<p>- Visualize the chemical structure of the final product in the Structure Explorer and download it in SVG format. Alternatively, copy the final structure's SMILES directly into ChemDraw or MarvinSketch for structure modification.</p>
<h2 id="alola_output">ALOLA Output</h2>
<p>- Once you've successfully modified and generated your desired Biosynthetic Gene Cluster (BGC) using ALOLA, you have the convenience of saving the output in various formats for future use. ALOLA provides users with flexibility by allowing the export of your customized BGC and the final chemical structure in three different file formats: PNG, SVG, and SMILIEs.</p>
<p><strong><em>PNG Format</em></strong></p>
<p>- The PNG format (Portable Network Graphics) is a widely supported image format. Saving your BGC in PNG ensures compatibility with various image viewing and editing tools. This raster image format preserves the visual details of your modified BGC, making it suitable for documentation, presentations, and further analysis.</p>
<p><strong><em>SVG Format</em></strong></p>
<p>- SVG, Scalable Vector Graphics, is a versatile and editable vector image format. Saving your modified BGC in SVG allows for easy scaling without loss of quality. This format is particularly useful if you plan to make additional adjustments or if you require a scalable image for publication or collaborative work.</p>
<p><strong><em>SMILIES Format</em></strong></p>
<p>- ALOLA offers the unique option to save your modified BGC in the SMILIES (Simplified Molecular Input Line Entry System) format. SMILIEs allows chemists to describe the structure of a molecule using a simple and standardized text string.</p>
</div>
</body>
</html>
